Add gradle build for the plugin

kandoeniko opened the pull request:
- configure the appveyor and travis
jobs

Commit:
Add gradle build for the plugin

- configure the appveyor and travis
jobs
- add empty plugin in order to have
the setup done correctly for the plugin

tephe commented on build.gradle:
```diff
@@ -0,0 +1,38 @@
+}
+
+group = 'com.ullink.gradle'
+description 'Uses sonar-scanner-msbuild to do static analysis with Sonarqube.'
```
Is this the description we want?

kandoeniko replied:
Nope, fixed it. Thanks!

tephe reviewed (commented):
I think this project should be named gradle-extentreports-csharp-plugin, not dotnet,  because it uses the extentreports-csharp project. Then it would be easier to find by others :)

kandoeniko commented:
Actually the reporting framework is called extentreports-dotnet-cli, so I think it is ok with the current naming (to avoid having the csharp part in the name)

Commit:
Generate report file based on TestResult

- use the extent-report framework
- generate report once NUnit or OpenCover
finished generating the report
- added unit tests for report generation

ngyukman commented on src/main/groovy/com/ullink/gradle/extentreport/ReportGenerator.groovy:
```diff
@@ -0,0 +1,88 @@
+class ReportGenerator extends Exec {
+
+    def EXTENT_REPORT_TYPE = 'v3html'
+    def EXTENT_REPORT_DOWNLOAD_URL = 'https://www.nuget.org/api/v2/package/extent/0.0.3'
```
can extract the version as separated property so project can upgrade to newer version easily?

kandoeniko replied:
I thought about it. But it actually defines just the style of the generated html.  I don't think we will actually configure this or have a lot of changes, so I let it be as it is for now.

ngyukman commented on src/main/groovy/com/ullink/gradle/extentreport/ReportGenerator.groovy:
```diff
@@ -0,0 +1,88 @@
+
+class ReportGenerator extends Exec {
+
+    def EXTENT_REPORT_TYPE = 'v3html'
```
property that can be configured should use camel case not upper cases

kandoeniko replied:
I thought that this was the convention for consts in groovy :)

ngyukman replied:
for constant yes, but it is now defined as a property of the task

kandoeniko replied:
Done

dfev77 commented on src/main/groovy/com/ullink/gradle/extentreport/ReportGenerator.groovy:
```diff
@@ -0,0 +1,88 @@
+            ensureExtentReportIsAvailable()
+            generateReports()
+        } else {
+            project.logger.info("There is no available Test Result file based on which the report should be generated located at ${project.tasks.nunit.getTestReporthPath()}")
```
put a wrapper method around project.tasks.nunit.getTestReportPath() as it is used in multiple places

dfev77 commented on src/main/groovy/com/ullink/gradle/extentreport/ReportGenerator.groovy:
```diff
@@ -0,0 +1,88 @@
+        return new File(extentReportFolder, EXTENT_REPORT_EXECUTABLE_PATH)
+    }
+
+    void downloadExtentReport() {
```
can we separate the download of extent in a separate plugin and have this plugin depend on that?
Just to keep things cleaner

ngyukman commented on build.gradle:
```diff
@@ -0,0 +1,45 @@
+    authorId 'ekando'
+    authorName 'Kando Eniko'
+    authorEmail 'eniko.kando@gmail.com'
+    projectUrl "https://github.com/Ullink/${project.name}"
```
it's Itiviti now :)
```suggestion
    projectUrl "https://github.com/Itiviti/${project.name}"
```

ngyukman commented on build.gradle:
```diff
@@ -0,0 +1,45 @@
+}
+
+plugindev {
+    pluginName 'com.ullink.gradle:gradle-extentreports-dotnet-plugin'
```
as a new plugin we should use the new group name com.itiviti

ngyukman commented on build.gradle:
```diff
@@ -0,0 +1,45 @@
+
+bintray {
+    user project.properties.bintrayUser
+    key project.properties.bintrayApiKey
```
from the plugindev [documentation](https://github.com/etiennestuder/gradle-plugindev-plugin#configure-bintray-plugin), bintray closure can be put after plugindev and it will assign user and key correctly

kandoeniko replied:
Ok, so if I understood correctly, I just need to move it after the plugindev and there is no need to specifically set the user & key ?

ngyukman replied:
yes, it should be the case

ngyukman commented on build.gradle:
```diff
@@ -0,0 +1,45 @@
+        [group: 'com.ullink.gradle', name: 'gradle-nunit-plugin', version: '1.14'],
+        [group: 'com.ullink.gradle', name: 'gradle-opencover-plugin', version: '1.13']
+    )
+    testCompile 'junit:junit:4.11'
```
don't think junit is needed to add explicitly when spock is used

ngyukman commented on gradle.properties:
```diff
@@ -0,0 +1 @@
+version=1.0
```
following semver is better, so it would be 1.0.0

ngyukman commented on src/main/groovy/com/ullink/gradle/extentreport/ExtentReportDotnetPlugin.groovy:
```diff
@@ -0,0 +1,28 @@
+    void apply(Project project) {
+
+        Task extentReportDownloaderTask = project.task('reportDownload', type: ReportGeneratorDownloader)
+        Task reportingTask = project.task('nunitReport', type: ReportGenerator)
```
should add task group and description to show the task from ./gradlew tasks [see here](https://guides.gradle.org/writing-gradle-tasks/#add_a_task_description)

ngyukman commented on src/main/groovy/com/ullink/gradle/extentreport/ReportGenerator.groovy:
```diff
@@ -0,0 +1,49 @@
+
+class ReportGenerator extends Exec {
+
+    def ExtentReportType = 'v3html'
```
naming should be camel case 
```suggestion
    def extentReportType = 'v3html'
```

ngyukman commented on src/main/groovy/com/ullink/gradle/extentreport/ReportGeneratorDownloader.groovy:
```diff
@@ -0,0 +1,57 @@
+
+class ReportGeneratorDownloader extends Download {
+
+    def EXTENT_REPORT_DOWNLOAD_URL = 'https://www.nuget.org/api/v2/package/extent/0.0.3'
```
it should be camel case as well
furthermore, usually we try to extract the version out so we can upgrade easily, like opencover or nunit

ngyukman commented on src/main/groovy/com/ullink/gradle/extentreport/ReportGeneratorDownloader.groovy:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    void downloadExtentReport() {
+        def downloadedFile = new File(getTemporaryDir(), EXTENT_REPORT_DOWNLOADED_FILE_NAME)
```
you can do getTemporaryDir().createNewFile() instead, as the name is not really necessary to be a constant / variable

ngyukman commented on src/test/groovy/com/ullink/ExtentReportDotnetPluginTest.groovy:
```diff
@@ -0,0 +1,55 @@
+import spock.lang.Specification
+
+class ExtentReportDotnetPluginTest extends Specification{
+    @Rule TemporaryFolder testProjectDir = new TemporaryFolder()
```
in spock, good thing is you don't need to new for the Rule
```suggestion
    @Rule TemporaryFolder testProjectDir
```
